# 1. EDA
    1. Data Profiling
    2. Stastical analysis
    3. Graphical Analysis

# 2. Feature Engineering
    1. Handling Missing values
    2. Encoding
    3. Transformation
    4. Handling outliers
    5. Feature Scaling

In [23]:
# importing required libraries for EDA
# the main aim is to understand data in better way
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statistics as stat
%matplotlib inline
# To display maximum columns of dataframe on screen
pd.pandas.set_option('display.max_columns', None)

# 1.0 Importing data set and gathering basic info about dataset

In [24]:
# importing dataset and displaying top 5 rows
dataset=pd.read_csv('Travel.csv')
dataset.head()

CustomerID  ProdTaken   Age    TypeofContact  CityTier  DurationOfPitch  \
0      200000          1  41.0     Self Enquiry         3              6.0   
1      200001          0  49.0  Company Invited         1             14.0   
2      200002          1  37.0     Self Enquiry         1              8.0   
3      200003          0  33.0  Company Invited         1              9.0   
4      200004          0   NaN     Self Enquiry         1              8.0   

       Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0        Salaried  Female                       3                3.0   
1        Salaried    Male                       3                4.0   
2     Free Lancer    Male                       3                4.0   
3        Salaried  Female                       2                3.0   
4  Small Business    Male                       2                3.0   

  ProductPitched  PreferredPropertyStar MaritalStatus  NumberOfTrips  \
0         Deluxe                    3.0        Single            1.0   
1         Deluxe                    4.0      Divorced            2.0   
2          Basic                    3.0        Single            7.0   
3          Basic                    3.0      Divorced            2.0   
4          Basic                    4.0      Divorced            1.0   

   Passport  PitchSatisfactionScore  OwnCar  NumberOfChildrenVisiting  \
0         1                       2       1                       0.0   
1         0                       3       1                       2.0   
2         1                       3       0                       0.0   
3         1                       5       1                       1.0   
4         0                       5       1                       0.0   

  Designation  MonthlyIncome  
0     Manager        20993.0  
1     Manager        20130.0  
2   Executive        17090.0  
3   Executive        17909.0  
4   Executive        18468.0

In [25]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   int64  
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   int64  
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4862 non-null   float64
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

### Observation
1. There is total 4888 rows and 20 columns.
2. There is total 3 data types float64, int64 and object.

## 1.1 Seperating numerical and categorical data

## 1.1.1 Numerical dataset

In [26]:
# 1. Getting Numerical features from dataset 
# 2. Creating Numerical dataframe
numerical_features=[feature for feature in dataset.columns if dataset[feature].dtypes!='O' and feature not in ['CustomerID']]
data_numerical=dataset[numerical_features]
data_numerical.head()

ProdTaken   Age  CityTier  DurationOfPitch  NumberOfPersonVisiting  \
0          1  41.0         3              6.0                       3   
1          0  49.0         1             14.0                       3   
2          1  37.0         1              8.0                       3   
3          0  33.0         1              9.0                       2   
4          0   NaN         1              8.0                       2   

   NumberOfFollowups  PreferredPropertyStar  NumberOfTrips  Passport  \
0                3.0                    3.0            1.0         1   
1                4.0                    4.0            2.0         0   
2                4.0                    3.0            7.0         1   
3                3.0                    3.0            2.0         1   
4                3.0                    4.0            1.0         0   

   PitchSatisfactionScore  OwnCar  NumberOfChildrenVisiting  MonthlyIncome  
0                       2       1                       0.0        20993.0  
1                       3       1                       2.0        20130.0  
2                       3       0                       0.0        17090.0  
3                       5       1                       1.0        17909.0  
4                       5       1                       0.0        18468.0

### 1.1.1.1 Discrete Numerical features

In [27]:
discrete_features=[feature for feature in numerical_features if len(data_numerical[feature].unique())<25 ]
print(discrete_features)
len(discrete_features)

['ProdTaken', 'CityTier', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'PreferredPropertyStar', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting']


10

### 1.1.1.2 Continuous Numerical features


In [28]:
continuous_features=[feature for feature in numerical_features if len(data_numerical[feature].unique())>25 ]
print(continuous_features)
len(continuous_features)

['Age', 'DurationOfPitch', 'MonthlyIncome']


3

## 1.1.2 Categorical dataset

In [29]:
# 1. Getting categorical features from dataset 
# 2. Creating categorical dataframe
categorical_features=[feature for feature in dataset.columns if dataset[feature].dtypes=='O' and feature not in ['CustomerID']]
data_categorical=dataset[categorical_features]
data_categorical.head()

TypeofContact      Occupation  Gender ProductPitched MaritalStatus  \
0     Self Enquiry        Salaried  Female         Deluxe        Single   
1  Company Invited        Salaried    Male         Deluxe      Divorced   
2     Self Enquiry     Free Lancer    Male          Basic        Single   
3  Company Invited        Salaried  Female          Basic      Divorced   
4     Self Enquiry  Small Business    Male          Basic      Divorced   

  Designation  
0     Manager  
1     Manager  
2   Executive  
3   Executive  
4   Executive

In [30]:
# Getting unique categories for each features in categorical features
for feature in categorical_features:
 print("The feature is '{}' and number of categories are '{}'".format(feature, len(dataset[feature].unique())))

The feature is 'TypeofContact' and number of categories are '3'
The feature is 'Occupation' and number of categories are '4'
The feature is 'Gender' and number of categories are '3'
The feature is 'ProductPitched' and number of categories are '5'
The feature is 'MaritalStatus' and number of categories are '4'
The feature is 'Designation' and number of categories are '5'


## 1.3 Misssing values

In [31]:
# getting features with missing values
feature_nan=[feature for feature in dataset.columns if dataset[feature].isnull().sum()>1 ]
print(feature_nan)

['Age', 'TypeofContact', 'DurationOfPitch', 'NumberOfFollowups', 'PreferredPropertyStar', 'NumberOfTrips', 'NumberOfChildrenVisiting', 'MonthlyIncome']


In [35]:
for feature in feature_nan:
    print(f"{feature} has {dataset[feature].isnull().sum()} null values")

Age has 226 null values
TypeofContact has 25 null values
DurationOfPitch has 251 null values
NumberOfFollowups has 45 null values
PreferredPropertyStar has 26 null values
NumberOfTrips has 140 null values
NumberOfChildrenVisiting has 66 null values
MonthlyIncome has 233 null values


In [36]:
# Getting percentage of missing values for individual features
for feature in feature_nan:
 print("{} has {} % of missing values".format(feature, np.round((dataset[feature].isnull().mean())*100,3)))

Age has 4.624 % of missing values
TypeofContact has 0.511 % of missing values
DurationOfPitch has 5.135 % of missing values
NumberOfFollowups has 0.921 % of missing values
PreferredPropertyStar has 0.532 % of missing values
NumberOfTrips has 2.864 % of missing values
NumberOfChildrenVisiting has 1.35 % of missing values
MonthlyIncome has 4.767 % of missing values


# 2.0 Statistical analysis
## 2.1 Numerical dataset
### 2.1.1 Mean, Median and Mode for numerical dataset

In [37]:
# mean of individual features
data_numerical.mean()

ProdTaken                       0.188216
Age                            37.622265
CityTier                        1.654255
DurationOfPitch                15.490835
NumberOfPersonVisiting          2.905074
NumberOfFollowups               3.708445
PreferredPropertyStar           3.581037
NumberOfTrips                   3.236521
Passport                        0.290917
PitchSatisfactionScore          3.078151
OwnCar                          0.620295
NumberOfChildrenVisiting        1.187267
MonthlyIncome               23619.853491
dtype: float64

In [39]:
# median of individual features
data_numerical.median()

ProdTaken                       0.0
Age                            36.0
CityTier                        1.0
DurationOfPitch                13.0
NumberOfPersonVisiting          3.0
NumberOfFollowups               4.0
PreferredPropertyStar           3.0
NumberOfTrips                   3.0
Passport                        0.0
PitchSatisfactionScore          3.0
OwnCar                          1.0
NumberOfChildrenVisiting        1.0
MonthlyIncome               22347.0
dtype: float64

In [40]:
# mode of individual features
data_numerical.mode().loc[0]

ProdTaken                       0.0
Age                            35.0
CityTier                        1.0
DurationOfPitch                 9.0
NumberOfPersonVisiting          3.0
NumberOfFollowups               4.0
PreferredPropertyStar           3.0
NumberOfTrips                   2.0
Passport                        0.0
PitchSatisfactionScore          3.0
OwnCar                          1.0
NumberOfChildrenVisiting        1.0
MonthlyIncome               17342.0
Name: 0, dtype: float64

In [41]:
# verifying modes
for feature in data_numerical.columns:
 print(feature,":",stat.mode(data_numerical[feature]))

ProdTaken : 0
Age : 35.0
CityTier : 1
DurationOfPitch : 9.0
NumberOfPersonVisiting : 3
NumberOfFollowups : 4.0
PreferredPropertyStar : 3.0
NumberOfTrips : 2.0
Passport : 0
PitchSatisfactionScore : 3
OwnCar : 1
NumberOfChildrenVisiting : 1.0
MonthlyIncome : 17342.0


## 2.1.2 Varience and Standard deviation for numerical dataset

In [42]:
# varience for individual features
round(data_numerical.var(),2)

ProdTaken                          0.15
Age                               86.80
CityTier                           0.84
DurationOfPitch                   72.58
NumberOfPersonVisiting             0.53
NumberOfFollowups                  1.01
PreferredPropertyStar              0.64
NumberOfTrips                      3.42
Passport                           0.21
PitchSatisfactionScore             1.87
OwnCar                             0.24
NumberOfChildrenVisiting           0.74
MonthlyIncome               28951914.85
dtype: float64

In [44]:
# Standard deviation for individual features
data_numerical.std()

ProdTaken                      0.390925
Age                            9.316387
CityTier                       0.916583
DurationOfPitch                8.519643
NumberOfPersonVisiting         0.724891
NumberOfFollowups              1.002509
PreferredPropertyStar          0.798009
NumberOfTrips                  1.849019
Passport                       0.454232
PitchSatisfactionScore         1.365792
OwnCar                         0.485363
NumberOfChildrenVisiting       0.857861
MonthlyIncome               5380.698361
dtype: float64